Add API stub for the USBEndpoint interface (#4)
GoogleChromeLabs/puppy-content: content/en-US/api/USBEndpoint/USBEndpoint.md

---
recipe: api-feature
title: 'USBEndpoint'
mdn_url: /en-US/docs/Web/API/USBEndpoint
specifications: https://wicg.github.io/webusb/#endpoints
browser_compatibility: api.USBEndpoint
---

## Description

> The `USBEndpoint` interface of the WebUSB API provides information about an endpoint provided by the USB device. An endpoint represents a unidirectional data stream into or out of the device. Endpoints come in three flavors,

Reviewer: Where's the three flavors?

Maybe say 'types' instead of 'flavors'. I would be ok with this word in a blog post, but for a technical reference it's a little colloquial.

Author: I decided to move the list of endpoint types to later in the document but forgot to remove this sentence fragment.

## Constructor

Creates a new `USBEndpoint` object which will be populated with information about the endpoint on the provided `USBAltenateInterface` with the given endpoint number and transfer direction.

## Properties

**`USBEndpoint.endpointNumber`**

Returns this endpoint's "endpoint number" which is a value from 1 to 15 extracted from the `bEndpointAddress` field of the endpoint descriptor defining this endpoint. This value is used to identify the endpoint when calling methods on `USBDevice`.

**`USBEndpoint.direction`**

> Returns the direction in which this endpoint transfers data. One of,

Reviewer: ```suggestion
Returns the direction in which this endpoint transfers data, one of:
```

> * `"in"` - data is transferred from device to host

Reviewer: ```suggestion
* `"in"` - Data is transferred from device to host.
```

> * `"out` - data is transferred from host to device

Reviewer: ```suggestion
* `"out"` - data is transferred from host to device.
```

**`USBEndpoint.type`**

> Returns the type of this endpoint. One of,

Reviewer: ```suggestion
Returns the type of this endpoint, one of:
```

> * `"bulk"` - Bulk endpoints provide reliable data transfer for large payloads. Data send through a bulk endpoint is guaranteed to be delivered or generate an error but may be preempted by other data traffic.

Reviewer: ```suggestion
* `"bulk"` - Provides reliable data transfer for large payloads. Data sent through a bulk endpoint is guaranteed to be delivered or generate an error but may be preempted by other data traffic.
```

Reviewer: Definition lists are one of the few places writers are allowed to use a sentence fragment. It eliminates duplicate information. Generally you start the fragment with a present tense verb.

> * `"interrupt"` - Interrupt endpoints provide reliable data transfer for small payloads. Data sent through an interrupt endpoint is guaranteed to be delivered or generate an error and is also given dedicated bus time for transmission.

Reviewer: ```suggestion
* `"interrupt"` - Provides reliable data transfer for small payloads. Data sent through an interrupt endpoint is guaranteed to be delivered or generate an error and is also given dedicated bus time for transmission.
```

> * `"isochronous'` - Isochronous endpoints provide unreliable data transfer for payloads that must be delivered periodically. They are given dedicated bus time but if a deadline is missed the data is dropped.

Reviewer: ```suggestion
* `"isochronous'` - Provides unreliable data transfer for payloads that must be delivered periodically. They are given dedicated bus time but if a deadline is missed the data is dropped.
```

**`USBEndpoint.packetSize`**

> Returns the size of the packets data sent through this endpoint will be divided into.

Reviewer: ```suggestion
Returns the size of the packets that data sent through this endpoint will be divided into.
```

## Examples

> While sometimes the developer knows ahead of time the exact layout of a device's endpoints there are cases where this must be discovered at runtime. For example, a USB serial device must provide bulk input and output endpoints but their endpoint numbers are not defined. This code will identify the correct endpoints.

Reviewer: Please add an opening sentence that explains generally what the example does.
